compiler/src/dotty/tools/dotc/inlines/Inlines.scala

```diff
@@ -345,10 +349,58 @@ object Inlines:
       val noRewriteSettings = ctx.settings.rewrite.updateIn(ctx.settingsState.reinitializedCopy(), None)
 
+      class MegaPhaseWithCustomPhaseId(miniPhases: Array[MiniPhase], startId: PhaseId, endId: PhaseId)
+        extends MegaPhase(miniPhases) {
+        override def start: Int = startId
+        override def end: Int = endId
+      }
+
       ConstFold(underlyingCodeArg).tpe.widenTermRefExpr match {
         case ConstantType(Constant(code: String)) =>
-          val source2 = SourceFile.virtual("tasty-reflect", code)
-          inContext(ctx.fresh.setSettings(noRewriteSettings).setNewTyperState().setTyper(new Typer(ctx.nestingLevel + 1)).setSource(source2)) {
+          val unitName = "tasty-reflect"
+          val source2 = SourceFile.virtual(unitName, code)
+          // We need a dummy owner, as the actual one does not have a computed denotation yet,
+          // but might be inspected in a transform phase, leading to cyclic errors
+          val dummyOwner = newSymbol(ctx.owner, "$dummySymbol$".toTermName, Private, defn.AnyType, NoSymbol)
+          val newContext =
+            ctx.fresh
+            .setSettings(noRewriteSettings)
+            .setNewTyperState()
+            .setTyper(new Typer(ctx.nestingLevel + 1))
+            .setSource(source2)
+            .withOwner(dummyOwner)
+
+          inContext(newContext) {
+            // Let's reconstruct necessary transform MegaPhases, without anything
+            // that could cause problems here (like `CrossVersionChecks`).
+            // The individiual lists here should line up with Compiler.scala, i.e
+            // separate chunks there should also be kept separate here.
+            // For now we create a single MegaPhase, since there does not seem to
+            // be any important checks later (e.g. ForwardDepChecks could be applicable here,
+            // but the equivalent is also not run in the scala 2's `ctx.typechecks`,
+            // so let's leave it out for now).
+            val transformPhases: List[List[(Class[?], () => MiniPhase)]] = List(
+              List(
+                (classOf[InlineVals], () => new InlineVals),
+                (classOf[ElimRepeated], () => new ElimRepeated),
+                (classOf[RefChecks], () => new RefChecks),
+              ),
+            )
+
+            val mergedTransformPhases =
+              transformPhases.flatMap( (megaPhaseList: List[(Class[?], () => MiniPhase)]) =>
+                val (newMegaPhasePhases, phaseIds) =
+                  megaPhaseList
+                    .flatMap { filteredPhase =>
+                      ctx.base.phases.find(phase => filteredPhase._1.isInstance(phase)).map { a =>
+                        (filteredPhase._2(), a.id)
+                      }
+                    }
+                    .unzip
+                if newMegaPhasePhases.isEmpty then None
+                else Some(MegaPhaseWithCustomPhaseId(newMegaPhasePhases.toArray, phaseIds.head, phaseIds.last))
+              )
```

Comment:
> Also, why not reuse the already existing MiniPhases here and just create the MegaPhase ?

Even though they are currently stateless, maybe it's best to create new phases in case one of them become stateful in the future. It will avoid having bugs that are hard to find.

Reply:
`MiniPhase`s do hold a reference to the outer MegaPhase (via `private[MegaPhase] var superPhase`), which is why they are recreated here. This `superPhase` field is assigned in a MegaPhase initialization, via a constructor call, so they are correctly initialized in this PR as well.